File: src/main/java/de/hsa/bpmnandai/project/domain/service/DroolsService.java
package de.hsa.bpmnandai.project.domain.service;

import de.hsa.bpmnandai.project.infrastructure.entity.BankLoanRequestEntity;
import de.hsa.bpmnandai.project.infrastructure.repository.BankLoanRequestRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.kie.dmn.api.core.*;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import java.util.Objects;

@Slf4j
@Service
@RequiredArgsConstructor
public class DroolsService {

    private final DMNRuntime dmnRuntime;
    private final BankLoanRequestRepository bankRepository;

    @Value("${dmn.drools.model.name}")
    private String dmnModelName;

    @Value("${dmn.drools.model.namespace}")
    private String dmnModelNameSpace;

    @Value("${dmn.drools.model.output}")
    private String dmnModelOutput;

    public String evaluateCreditApprovalDecisionModel(final BankLoanRequestEntity request) {
        final DMNModel dmnModel = initializeModel();
        final DMNDecisionResult isApproved = executeDecisionModel(dmnModel, request);
        return mapResult(isApproved);
    }

    private DMNModel initializeModel() {
        return dmnRuntime.getModel(dmnModelNameSpace, dmnModelName);
    }

    private DMNDecisionResult executeDecisionModel(final DMNModel dmnModel, final BankLoanRequestEntity request) {
        final DMNContext context = initializeContext(request);
        final DMNResult dmnResult = dmnRuntime.evaluateAll(dmnModel, context);
        return dmnResult.getDecisionResultByName(dmnModelOutput);
    }

    private String mapResult(final DMNDecisionResult result) {
        final Object rawResult = result.getResult();
        if (Objects.isNull(rawResult)) {
            return "UNKNOWN";
        } else {
            return (String) rawResult;
        }
    }

    private DMNContext initializeContext(final BankLoanRequestEntity entity) {
        final DMNContext dmnContext = dmnRuntime.newContext();
        dmnContext.set("Kreditanfrage", entity);
        return dmnContext;
    }

}

File: src/main/kotlin/de/emaarco/bpmnandai/project/domain/service/DroolsService.kt
package de.emaarco.bpmnandai.project.domain.service

import de.emaarco.bpmnandai.project.infrastructure.entity.BankLoanRequestEntity
import org.kie.dmn.api.core.*
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import java.util.*

@Service
class DroolsService(private val dmnRuntime: DMNRuntime) {

    @Value("\${dmn.drools.model.name}")
    private lateinit var dmnModelName: String

    @Value("\${dmn.drools.model.namespace}")
    private lateinit var dmnModelNameSpace: String

    @Value("\${dmn.drools.model.output}")
    private val dmnModelOutput: String? = null

    fun evaluateCreditApprovalDecisionModel(request: BankLoanRequestEntity): String {
        val dmnModel: DMNModel = initializeModel()
        val isApproved: DMNDecisionResult = executeDecisionModel(dmnModel, request)
        return mapResult(isApproved)
    }

    private fun initializeModel(): DMNModel {
        return dmnRuntime.getModel(dmnModelNameSpace, dmnModelName)
    }

    private fun executeDecisionModel(dmnModel: DMNModel, request: BankLoanRequestEntity): DMNDecisionResult {
        val context: DMNContext = initializeContext(request)
        val dmnResult: DMNResult = dmnRuntime.evaluateAll(dmnModel, context)
        return dmnResult.getDecisionResultByName(dmnModelOutput)
    }

    private fun mapResult(result: DMNDecisionResult): String {
        val rawResult: Any = result.result
        return if (Objects.isNull(rawResult)) {
            "UNKNOWN"
        } else {
            rawResult as String
        }
    }

    private fun initializeContext(entity: BankLoanRequestEntity): DMNContext {
        val dmnContext: DMNContext = dmnRuntime.newContext()
        dmnContext.set("Kreditanfrage", entity)
        return dmnContext
    }

}